### PR TITLE
Implementation of #129: Allow more tuning of the model hyperparameters. 

### LINKED ISSUE #129: Allow more tuning of the model hyperparameters.
# Allow more tuning of the model hyperparameters.

While working on #126, I noticed that in the constructor of LogisticRegressionModel, a few hyperparameters are hard-coded:

```
def __init__(self):
    self.logit = LogisticRegression(fit_intercept=True, C=1e9,
                                    solver='liblinear', random_state=42)     
```

When trying to optimize for the best model performance, one might want to try tuning the regularization hyperparameter C.
In extreme cases, one might want to set fit_intercept to False as well...
Also, in text-book approaches, even trying different random_states is advised, to counter against lucky shot models or to verify model variance.
And in case speed is of the essence, one might want to play around with different solvers.

This is now impossible. With the intent on a certain company case to maximally tune, I thought it could be a good idea to make this a possibility.

## Task Description

To do so, the constructor of the LogisticRegressionModel must be adapted to take overrides of the default arguments, if the data scientist wants to pass that.
It might be good to do this too for LinearRegressionModel, again for fit_intercept for example:

```
def __init__(self, *, fit_intercept=True, normalize=False, copy_X=True,
                 n_jobs=None, positive=False):
```

On the other hand, the case for normalize, n_jobs and positive is weak though, since we take care of that already internally in Cobra:
- normalize: our preprocessing takes care of the features, not sure if normalizing upfront would have effect, but one might try, if we allow it now..
- n_jobs: only has an effect when having multiple targets per the scikit-learn docs, we don't support that yet in Cobra
- positive: we take care of that in Cobra already.
But let's not "forbid" any tuning and make it possible for this as well. Even if it's just to allow someone to block the use of fit_intercept.

### PR DESCRIPTION
**Pull request by sandervh14**
====> **Do not merge yet! See below for explanation.** <=====

# Story Title

[#129 Allow more tuning of the model hyperparameters.](https://github.com/PythonPredictions/cobra/pull/129)

## Changes made

Altered the constructor of both LogisticRegressionModel and LinearRegressionModel to support override of the default hyperparameters, via kwargs passing.
Documented in the docstring of both models.
See the issue description for more extensive explanation of why one would want this.

** Do not merge this yet with the Development branch. ** Unit tests have not yet been adapted yet. Was a speedy commit of an idea I was playing with in a notebook, when investigating how to further tune the models for a client. I'll take the time next week to adapt the unit tests.

## How does the solution address the problem

This PR allows a data scientist to pass argument overrides for the constructor of the scikit-learn model used behind the scenes in Cobra, to try to tune further on a model's performance, if possible vs. Cobra's default settings for the model.

## Linked issues

Resolves #129

### CODE REVIEW

**Comment by sborms** [comment]
Looks good Sander! Code is clear to me.

Two remarks:
- We have a `get_intercept()` function in the model classes - what will the result be if no intercept is fitted? This is a barely used helper function, but better to show a graceful error if there is no intercept to retrieve.
- Agree that for linear models the extra parameters are already taken care of by Cobra, but let's allow for the flexibility. Maybe add a caveat in one of the classes (so at least we have it written down somewhere).

File: cobra/model_building/models.py [hunk 1]
```diff
@@ -258,10 +274,27 @@ class LinearRegressionModel:
...
+    def __init__(self, **kwargs):
+        # Initialize a scikit-learn linear regression model,
+        # with custom arguments passed by the data scientist (if any),
+        # supplemented with Cobra's default arguments, if a custom value was
+        # not provided by the data scientist for overriding purposes:
+        default_kwargs = dict(fit_intercept=True)
+        for kwarg, val in default_kwargs.items():
+            if kwarg not in kwargs:
+                kwargs[kwarg] = val
+        self.linear = LinearRegression(**kwargs)
```
Comment (ZlaTanskY) [inline cobra/model_building/models.py:296]: This `for` loop can be replaced by a oneliner by using `.update()` (see https://www.programiz.com/python-programming/methods/dictionary/update). In that case, `default_kwargs` best to be renamed to something like `model_kwargs`.
```suggestion
        model_kwargs = dict(fit_intercept=True)
        model_kwargs.update(kwargs)
        self.linear = LinearRegression(**model_kwargs)
```

Comment (sandervh14) [reply cobra/model_building/models.py:296]: Nice Jano, I was going to commit it straight away with the github interface, but then it isn't applied to LogisitcRegression as well. I'll have a look at it this afternoon and will include Sam's remarks

**Comment by sandervh14** [comment]
Hey Sam: 

> * Agree that for linear models the extra parameters are already taken care of by Cobra, but let's allow for the flexibility. 

You mean only linear regression right? Do you agree with me that for logistic regression it's certainly a good idea? (most of all for tuning regularization C?)

File: cobra/model_building/models.py (hunk 1 above)
Comment (ZlaTanskY) [reply cobra/model_building/models.py:296]: An idea to resolve duplicate code (and documentation to some extent) is to create a BaseModel class from which both the LinearRegression and LogisticRegression class inherit. 
But this is of course out of scope for this PR and should be considered in a separate PR.

**Comment by sborms** [comment]
> Hey Sam:
> 
> > * Agree that for linear models the extra parameters are already taken care of by Cobra, but let's allow for the flexibility.
> 
> You mean only linear regression right? Do you agree with me that for logistic regression it's certainly a good idea? (most of all for tuning regularization C?)

Indeed, linear regression. It's definitely a good idea, it's fine to have the flexibility for both logistic & linear regression.

File: cobra/model_building/models.py (hunk 1 above)
Comment (sandervh14) [reply cobra/model_building/models.py:296]: > An idea to resolve duplicate code (and documentation to some extent) is to create a BaseModel class from which both the LinearRegression and LogisticRegression class inherit.

I remembered this comment while fixing #126, I also found there was quite some duplication and I had time to do this, so I've done this today with the solution to #126, yippee! :-). For details, see explanation in https://github.com/PythonPredictions/cobra/pull/128#issuecomment-1143567992.

> But this is of course out of scope for this PR and should be considered in a separate PR.

I took the liberty to include the superclassing abstraction in #126 anyway, instead of a new issue & PR dedicated to it, since the evaluate() which I was fixing unit tests for, takes up a BIG chunk of the code of both LinearRegression and LogisticRegressionModel.
